udata_piwik/metrics.py

```diff
@@ -1,123 +1,170 @@
-    name = 'reuse_views'
-    display_name = _('Reuses views')
-
-    def get_value(self):
-        ids = [d.id for d in
-               (Reuse.objects(organization=self.target).only('id') or [])]
-        return int(Metrics.objects(object_id__in=ids, level='daily')
-                          .sum('values.nb_uniq_visitors'))
-
-
-def aggregate_datasets_daily(org, day):
-    keys = ['datasets_{0}'.format(k) for k in KEYS]
-    ids = [d.id for d in Dataset.objects(organization=org).only('id')]
-    metrics = Metrics.objects(object_id__in=ids,
-                              level='daily', date=day.isoformat())
-    values = [int(metrics.sum('values.{0}'.format(k))) for k in KEYS]
-    return Metrics.objects.update_daily(org, day, **dict(zip(keys, values)))
-
-
-def aggregate_reuses_daily(org, day):
-    keys = ['reuses_{0}'.format(k) for k in KEYS]
-    ids = [r.id for r in Reuse.objects(organization=org).only('id')]
-    metrics = Metrics.objects(object_id__in=ids,
-                              level='daily', date=day.isoformat())
-    values = [int(metrics.sum('values.{0}'.format(k))) for k in KEYS]
-    Metrics.objects.update_daily(org, day, **dict(zip(keys, values)))
-
-
-def upsert_metric_for_day(obj, day, data):
-    oid = obj.id if hasattr(obj, 'id') else obj
-    if not isinstance(day, str):
-        day = (day or date.today()).isoformat()
-    commands = dict(('inc__values__{0}'.format(k), data[k]) for k in KEYS)
-    metrics = Metrics.objects(object_id=oid, level='daily', date=day)
-    return metrics.update_one(upsert=True, **commands)
-
-
-def clear_metrics_for_day(day):
-    if not isinstance(day, str):
-        day = (day or date.today()).isoformat()
-    commands = dict(('unset__values__{0}'.format(k), 1) for k in KEYS)
-    metrics = Metrics.objects(level='daily', date=day)
-    return metrics.update(upsert=False, **commands)
+def update_metrics_from_backend():
+    '''
+    TODO: factorize those fns (maybe class based?)
+    '''
+    update_resources_metrics_from_backend()
+    update_datasets_metrics_from_backend()
+
+
+def update_resources_metrics_from_backend():
+    '''
+    Update resources' metrics from backend.
+
+    Get a sum of all metrics for `resource_views` on the backend and
+    attach them to `resource.metrics`.
+    '''
+    log.info('Updating resources metrics from backend...')
+    client = metrics_client_factory()
+    result = client.get_views_from_all_datasets()
+    for (_, keys), _values in result.items():
+        values = next(_values)
+        values.pop('time')
+        resource_id = keys['resource']
+        try:
+            resource_id = uuid.UUID(resource_id)
+            resource = get_resource(resource_id)
+        except Exception as e:
+            log.exception(e)
+            continue
+        if resource:
+            log.debug('Found resource %s: %s', resource.id, values)
+            resource.metrics.update(values)
+            try:
+                # TODO: disable signals
```

Comment:
@quaxsze this is important: if we have a signal in `udata` that triggers on resource/dataset modification, we must not trigger it from here. Same for all objects.